gnosis/safe/account_abstraction/safe_operation.py

```diff
@@ -0,0 +1,140 @@
+import dataclasses
+from typing import Optional
+
+from eth_abi import encode as abi_encode
+from eth_abi.packed import encode_packed
+from eth_typing import ChecksumAddress
+from hexbytes import HexBytes
+
+from gnosis.eth.account_abstraction import UserOperation
+from gnosis.eth.utils import fast_keccak
+
+_domain_separator_cache = {}
+
+
+@dataclasses.dataclass
+class SafeOperation:
+    """
+    Safe EIP4337 operation
+    More info: https://github.com/safe-global/safe-modules/blob/main/modules/4337/contracts/Safe4337Module.sol#L55
+
+    TypeHash calculation:
+
+    .. code-block:: python
+
+        keccak256(
+            "SafeOp(address safe,uint256 nonce,bytes initCode,bytes callData,uint256 callGasLimit,uint256 verificationGasLimit,uint256 preVerificationGas,uint256 maxFeePerGas,uint256 maxPriorityFeePerGas,bytes paymasterAndData,uint48 validAfter,uint48 validUntil,address entryPoint)"
+        )
+
+    Domain separator calculation:
+
+    .. code-block:: python
+        keccak256("EIP712Domain(uint256 chainId,address verifyingContract)"
+
+    """
+
+    safe: ChecksumAddress
+    nonce: int  # uint256
+    init_code_hash: bytes  # bytes32
+    call_data_hash: bytes  # bytes32
+    call_gas_limit: int  # uint256
+    verification_gas_limit: int  # uint256
+    pre_verification_gas: int  # uint256
+    max_fee_per_gas: int  # uint256
+    max_priority_fee_per_gas: int  # uint256
+    paymaster_and_data_hash: bytes  # bytes32
+    valid_after: int  # uint48
+    valid_until: int  # uint48
+    entry_point: ChecksumAddress
+    signature: bytes
+    TYPE_HASH: bytes = HexBytes(
+        "0x84aa190356f56b8c87825f54884392a9907c23ee0f8e1ea86336b763faf021bd"
+    )  # bytes32
+    DOMAIN_SEPARATOR_TYPE_HASH: bytes = HexBytes(
+        "0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218"
+    )  # bytes32
+    safe_operation_hash: Optional[bytes] = None
+
+    @classmethod
+    def from_user_operation(cls, user_operation: UserOperation):
```

Comment:
With class extension this function should be the __init__ and you can add the hashes as cached properties.